src/test/java/org/broadinstitute/hellbender/utils/downsampling/PositionalDownsamplerUnitTest.java

```diff
@@ -0,0 +1,207 @@
+
+        Utils.resetRandomGenerator();
+        downsampler.submit(allReads);
+
+        final boolean expectFinalizedItemsAfterSubmission = expectedStackSizes.size() > 1 || (expectedStackSizes.size() == 1 && lastStackIsPureUnmapped);
+        Assert.assertEquals(downsampler.hasFinalizedItems(), expectFinalizedItemsAfterSubmission);
+        if ( expectFinalizedItemsAfterSubmission ) {
+            Assert.assertNotNull(downsampler.peekFinalized());
+        }
+        else {
+            Assert.assertNull(downsampler.peekFinalized());
+        }
+
+        final boolean expectPendingItemsAfterSubmission = ! expectedStackSizes.isEmpty() && ! lastStackIsPureUnmapped;
+        Assert.assertEquals(downsampler.hasPendingItems(), expectPendingItemsAfterSubmission);
+        if ( expectPendingItemsAfterSubmission ) {
+            Assert.assertNotNull(downsampler.peekPending());
+        }
+        else {
+            Assert.assertNull(downsampler.peekPending());
+        }
+
+        downsampler.signalEndOfInput();
+
+        final boolean expectFinalizedItemsAfterEndOfInput = ! expectedStackSizes.isEmpty();
+        Assert.assertEquals(downsampler.hasFinalizedItems(), expectFinalizedItemsAfterEndOfInput);
+        if ( expectFinalizedItemsAfterEndOfInput ) {
+            Assert.assertNotNull(downsampler.peekFinalized());
+        }
+        else {
+            Assert.assertNull(downsampler.peekFinalized());
+        }
+
+        Assert.assertFalse(downsampler.hasPendingItems());
+        Assert.assertNull(downsampler.peekPending());
+
+        List<GATKRead> downsampledReads = downsampler.consumeFinalizedItems();
+        Assert.assertFalse(downsampler.hasFinalizedItems() || downsampler.hasPendingItems());
+        Assert.assertTrue(downsampler.peekFinalized() == null && downsampler.peekPending() == null);
+
+        if ( expectedStackSizes.size() == 0 ) {
+            Assert.assertTrue(downsampledReads.isEmpty());
+        }
+        else {
+            List<Integer> downsampledStackSizes = getDownsampledStackSizesAndVerifySortedness(downsampledReads);
+
+            Assert.assertEquals(downsampledStackSizes.size(), expectedStackSizes.size());
+            Assert.assertEquals(downsampledStackSizes, expectedStackSizes);
+
+            int numReadsActuallyEliminated = allReads.size() - downsampledReads.size();
+            int numReadsReportedEliminated = downsampler.getNumberOfDiscardedItems();
+            Assert.assertEquals(numReadsActuallyEliminated, numReadsReportedEliminated);
+        }
+
+        downsampler.resetStats();
+        Assert.assertEquals(downsampler.getNumberOfDiscardedItems(), 0);
+
+        // Now test with a PositionalDownsampler wrapped in an iterator, and make sure we get the same results.
+        // It's crucial to reset the random number generator again in order to match the selections made by the
+        // first downsampling pass.
+        Utils.resetRandomGenerator();
```

Comment:
We want to assert that we get the exact same reads from the iterator as from using the downsampler directly, so we can't split it into a separate test. I think making it a separate non-test method would increase confusion.